common/sampling.cpp

```diff
@@ -167,9 +167,13 @@ llama_token llama_sampling_sample(
         llama_sample_grammar(ctx_main, &cur_p, ctx_sampling->grammar);
     }
 
-    if (temp <= 0) {
-        // greedy sampling
+    if (temp < 0.0) {
```

Comment:
The result is the same in either case right? I'm not entirely sure it's worth special casing this instead of just changing greedy sampling to do:

```c++
        llama_sample_softmax(ctx_main, &cur_p);
        id = cur_p.data[0].id;
```

But if you did go that way, you'd probably also want to change the common args parsing stuff to clamp the user-specified temperature to `0.0` so if they pass a negative value it's the same.

It's only internal stuff that would care about probs generated vs no probs unless I'm misunderstanding.

Reply:
It's the same result yes. The probs are not used only internally - we are using them in `speculative`. Before this PR, we had to do the hack with `temp = 0.01f;` to get probs. Now we get them with `temp = 0.0f;`

The user specified input should not be affected by this change. Technically, the user would normally want to pass `temp = -1.0f` to save the extra softmax compute, but it's probably not something that would affect performance in measurable way.

Reply:
Sorry, "internal" was a poor choice of words. I meant it's not something someone calling the application and passing `--temp` on the commandline would care about. So if they do `--temp -1` for an example that doesn't care about probs then it's kind of weird/unnecessary to turn on generating probs in that case.

So what I'm proposing is that the argument handling stuff would do something like:

```c++
params.sparams.temp = std::max(0.0f, atof(blah));
```

when parsing the commandline arguments, so even if the user does `--temp -1` it's still just `0.0`. Then something like `speculative` which cares about probs in the greedy sampling case can do:

```c++
if (params.sparams.temp == 0.0f) {
    params.sparams.temp = -1.0f;
}
```

Reply:
Got it. Should be good now